### PR TITLE
vpd-tool:Option to fixSystemVPD keywords

### PR DESCRIPTION
**Pull request by GiridhariKrishnan**
This commit has an added option in vpd-tool that fixes system VPD restorable keywords in an interactive way.
This option is required in a field usecase.

In the field when there is a fault in BMC and BMC gets replaced, the replaced BMC VPD will now have all blank data. Now the Product engineer can use vpd-tool fixSystemVPD option to update the BMC critical keywords by chosing the corresponding keyword data from system backplane.

Similarly when the system backplane is replaced,
using the option, the PEs' can copy the keyword data from BMC.

When both system backplane and BMC are replaced, the PEs' will have option to update new value on both.

This interactive tool option eases keyword data updation and avoids manual updation for each and every mismatching critical keyword.

Test: Tested on everest

// Create a mismatch.
root@ever8bmc:/tmp# ./vpd-tool-writeHardwareFix -w -H -O /sys/bus/i2c/drivers/at24/8-0050/eeprom -R VCEN -K FC -V 0x64560204 Disabled reboot
Sleep started
Enabled reboot
root@ever8bmc:/tmp#
root@ever8bmc:/tmp# busctl set-property xyz.openbmc_project.Inventory.Manager /xyz/openbmc_project/inventory/system/chassis/motherboard com.ibm.ipzvpd.LXR0 LX ay 2 0x04 0x64 root@ever8bmc:/tmp#
root@ever8bmc:/tmp# busctl set-property xyz.openbmc_project.Inventory.Manager /xyz/openbmc_project/inventory/system/chassis/motherboard com.ibm.ipzvpd.VSYS WN ay 3 0x74 0x85 0x69 root@ever8bmc:/tmp# ./vpd-tool-writeHardwareFix -w -H -O /sys/bus/i2c/drivers/at24/8-0050/eeprom -R VSYS -K WN -V 0x64560204077675 Disabled reboot
Sleep started
Enabled reboot

=========

(Trimmed the output)
root@ever8bmc:/tmp# ./vpd-tool --fixSystemVPD

Restorable record-keyword pairs and their data on cache & hardware.

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
|======================================================================================================|
1     UTIL    D0       0x69                          0x69                          NO
|------------------------------------------------------------------------------------------------------|
2     LXR0    LX       0x0464                        0x3100080100300074            YES
|------------------------------------------------------------------------------------------------------|
3     VCEN    FC       0x373830432d303031            0x645602042d303031            YES
|------------------------------------------------------------------------------------------------------|
4     VCEN    SE       RCH0014                       RCH0014                       NO
|------------------------------------------------------------------------------------------------------|
5     VSYS    BR       0x5330                        0x5330                        NO
|------------------------------------------------------------------------------------------------------|
6     VSYS    TM       0x646534302d4d5258            0x646534302d4d5258            NO
|------------------------------------------------------------------------------------------------------|
7     VSYS    SE       13E8CEX                       13E8CEX                       NO
|------------------------------------------------------------------------------------------------------|
8     VSYS    SU       0x0004ac1e9fd4                0x0004ac1e9fd4                NO
|------------------------------------------------------------------------------------------------------|
9     VSYS    RB       0x31000000                    0x31000000                    NO
|------------------------------------------------------------------------------------------------------|
10    VSYS    WN       0x748569                      0x645602040776754243453432    YES
|------------------------------------------------------------------------------------------------------|
11    VSYS    RG       0x00000000                    0x00000000                    NO
|------------------------------------------------------------------------------------------------------|

Enter 1 => If you choose cache data of all mismatching record-keyword pairs Enter 2 => If you choose hardware data of all mismatching record-keyword pairs Enter 3 => If you wish to explore more options
Enter 0 to exit  3

|======================================================================================================|

Iterate through all restorable record-keyword pairs

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
1     UTIL    D0       0x69                          0x69                          NO
|======================================================================================================|

No mismatch found.

Enter 6 => If you wish to enter a new value to update both on BMC and System Backplane Enter 7 => If you wish to skip the above record-keyword pair Enter 0 => To exit successfully : 6

|======================================================================================================|

Enter the new value to update both on BMC & System Backplane (Value should be in ASCII or in HEX(prefixed with 0x)) : 0x02

|======================================================================================================|

Data updated successfully.

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
2     LXR0    LX       0x0464                        0x3100080100300074            YES
|======================================================================================================|

Enter 4 => If you choose the cache value as the right value Enter 5 => If you choose hardware value as the right value Enter 6 => If you wish to enter a new value to update both cache and hardware Enter 7 => If you wish to skip the above record-keyword pair Enter 0 => To exit successfully  5

|=======================================================================| Disabled reboot
Sleep started, try to reboot
Sleep end
Sleep started
Enabled reboot

Data updated successfully.

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
3     VCEN    FC       0x373830432d303031            0x645602042d303031            YES
|======================================================================================================|

Enter 4 => If you choose the cache value as the right value Enter 5 => If you choose hardware value as the right value Enter 6 => If you wish to enter a new value to update both cache and hardware Enter 7 => If you wish to skip the above record-keyword pair Enter 0 => To exit successfully  4

|=======================================================================| Disabled reboot
Sleep started, try to reboot
Sleep end
Sleep started
Enabled reboot

Data updated successfully.

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
4     VCEN    SE       RCH0014                       RCH0014                       NO
|======================================================================================================|

No mismatch found.

Enter 6 => If you wish to enter a new value to update both on BMC and System Backplane Enter 7 => If you wish to skip the above record-keyword pair Enter 0 => To exit successfully : 7

|======================================================================================================|

Skipped the above record-keyword pair. Continueing to the next pair if available.

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
10    VSYS    WN       0x748569                      0x645602040776754243453432    YES
|======================================================================================================|

Enter 4 => If you choose the cache value as the right value Enter 5 => If you choose hardware value as the right value Enter 6 => If you wish to enter a new value to update both cache and hardware Enter 7 => If you wish to skip the above record-keyword pair Enter 0 => To exit successfully  6

|=======================================================================|

Enter the new value to update both in cache & hardware (Value should be in ASCII or in HEX(prefixed with 0x)) : 0x433035303736304243453432

|=======================================================================|

Data updated successfully.

|======================================================================================================|
S.No  Record  Keyword  Data On Cache                 Data On Hardware              Data Mismatch
11    VSYS    RG       0x00000000                    0x00000000                    NO
|======================================================================================================|
No mismatch found. Skip this record-keyword pair.

Signed-off-by: Priyanga Ramasamy <priyanga24@in.ibm.com>
Change-Id: Ifb32efd1174360409b795b162282673a4f73df1b

Conflicts:
	ibm_vpd_utils.hpp

### CODE REVIEW

**Review by PriyangaRamasamy — APPROVED** [review]

File: ibm_vpd_utils.hpp [hunk 1]
```diff
@@ -77,6 +77,68 @@ string encodeKeyword(const string& kw, const string& encoding);
...
+/** @brief A templated function to read D-Bus properties.
+ *
+ *  @param[in] service - Service path
+ *  @param[in] object - object path
+ *  @param[in] inf - interface
+ *  @param[in] prop - property whose value is fetched
+ *  @return The property value of its own type.
+ */
+template <typename T>
+T readDBusProperty(const string& service, const string& object,
```
Comment (SunnySrivastava1984) [inline ibm_vpd_utils.hpp:89]: Not sure why we have duplicate function for this.

File: vpd_tool_impl.cpp [hunk 2]
```diff
@@ -574,3 +622,409 @@ void VpdTool::readKwFromHw()
...
+        cout << '\n' << outline << endl;
+
+        if (json.find("frus") == json.end())
+        {
+            throw runtime_error("Frus not found in json");
+        }
+
+        bool mismatchFound = false;
+
+        if (option == VpdTool::BMC_DATA_FOR_ALL)
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.cpp:868]: Instead of if/else of switch case?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: vpd_tool_impl.cpp [hunk 3]
```diff
@@ -574,3 +622,411 @@ void VpdTool::readKwFromHw()
...
+                            printFixSystemVPDOption(VpdTool::EXIT);
+                        }
+
+                        cin >> option;
+                        cout << '\n' << outline << endl;
+
+                        EditorImpl edit(constants::systemVpdFilePath, json,
+                                        get<1>(data), get<2>(data));
+
+                        if (option == VpdTool::BMC_DATA_FOR_CURRENT)
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.cpp:976]: Why not switch for this as well?

File: vpd_tool_impl.cpp [hunk 4]
```diff
@@ -574,3 +622,411 @@ void VpdTool::readKwFromHw()
...
+                        }
+                    } while (1);
+                }
+                exit(0);
+            case VpdTool::EXIT:
+                cout << "\nExit successfully";
+                exit(0);
+            default:
+                cout << "\nProvide a valid option. Retry.";
+                continue;
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.cpp:1029]: Is continue required here?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: vpd_tool_impl.cpp (hunk 4 above)
Comment (GiridhariKrishna) [reply vpd_tool_impl.cpp:1029]: Yes, as it has to loop for retry.

Comment (GiridhariKrishna) [reply vpd_tool_impl.cpp:1029]: Can be removed

File: vpd_tool_impl.cpp [hunk 5]
```diff
@@ -18,6 +19,53 @@ namespace fs = std::filesystem;
...
+    auto parseResult = parser->parse();
+    ParserFactory::freeParser(parser);
+
+    if (auto pVal = std::get_if<Store>(&parseResult))
+    {
+        vpdMap = pVal->getVpdMap();
+    }
+    else
+    {
+        std::string err =
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.cpp:64]: Unnecessary copy. Please create the message in place.

File: vpd_tool_impl.cpp [hunk 6]
```diff
@@ -574,3 +622,405 @@ void VpdTool::readKwFromHw()
...
+    {
+        throw runtime_error("VPD JSON file not found");
+    }
+    js = json::parse(inventoryJson);
+
+    vpdVector = getVpdDataInVector(js, constants::systemVpdFilePath);
+    ParserInterface* parser = ParserFactory::getParser(
+        vpdVector, constants::pimPath + std::string(constants::SYSTEM_OBJECT));
+    auto parseResult = parser->parse();
+    ParserFactory::freeParser(parser);
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.cpp:697]: Can we use the function defined above? "getVPDInMap"

File: vpd_tool_impl.hpp [hunk 7]
```diff
@@ -9,6 +9,11 @@
 
 using json = nlohmann::json;
 
+// <S.no, Record, Keyword, D-Bus value, HW value, Data mismatch>
+using SystemCriticalData =
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.hpp:13]: Define this in types.hpp.

File: vpd_tool_impl.hpp [hunk 8]
```diff
@@ -190,6 +232,16 @@ class VpdTool
...
 
+    /**
+     * @brief Fix System VPD keyword.
+     * This API provides an interactive way to fix system VPD keywords that are
+     * part of restorable record-keyword pairs. The user can use this option to
+     * restore the restorable keywords in cache or in hardware or in both cache
+     * and hardware.
+     * @return returncode (success/failure).
+     */
+    int fixSystemVPD();
```
Comment (SunnySrivastava1984) [inline vpd_tool_impl.hpp:243]: Why int is required to return success/failure?

**Review by SunnySrivastava1984 — CHANGES_REQUESTED** [review]

File: vpd_tool_impl.cpp (hunk 6 above)
Comment (GiridhariKrishna) [reply vpd_tool_impl.cpp:697]: Yes, but as it's cherry-pick change so didn't want to change in lower release.

File: vpd_tool_impl.cpp (hunk 5 above)
Comment (GiridhariKrishna) [reply vpd_tool_impl.cpp:64]: yes, okay.

File: ibm_vpd_utils.hpp (hunk 1 above)
Comment (GiridhariKrishna) [reply ibm_vpd_utils.hpp:89]: As discussed before these are cherry-pick change, and planned not to change for 1020.

File: vpd_tool_impl.hpp (hunk 8 above)
Comment (GiridhariKrishna) [reply vpd_tool_impl.hpp:243]: These are from 1030 and are in 1030,1050

File: vpd_tool_impl.hpp (hunk 7 above)
Comment (GiridhariKrishna) [reply vpd_tool_impl.hpp:13]: okay.

**Review by SunnySrivastava1984 — APPROVED** [review]
As these are cherry-pick changes from future release branch. Does not need modification here.

**Pull request merged**